### ISSUE TITLE
Weighted queues

### ISSUE THREAD

**Issue by dkam** [beanstalkd/beaneater#85]
Hi there,
    First, thanks for writing this excellent software - I've been using it for 10+ years and processed literally billions of jobs with it.

I recently started using weighted queues in another project which uses Sidekiq and I really enjoy not having to worry about fine turning priorities and running multiple job processors to avoid queue starvation. It's nice to get a predictable amount of processing for every queue.

Is there any way to achieve something similar with Beanstalkd? Essentially, selecting from job queues in a weighted random fashion?

I imagine I could rig up a system that uses weights to randomly select a queue, peak to see if there are jobs ready and if so, reserve a job from that queue.  Rough code below:

```ruby
pipes = [{pipe: 'low_priority', weight: 1}, {pipe: 'medium_priority', weight: 2}, {pipe: 'high_priority', weight: 4}]

q_min = pipes.min_by {|q| q.dig(:weight)}.dig(:weight)
q_max = pipes.inject(0) {|r, q| r + q.dig(:weight) }
range = q_min..q_max

loop do
  ## Randomly select a queue based on weights and assign queue name to pipe_to_process

  q_rand = Random.rand(range)
  q_accumulate = 0
  pipe_to_process = pipes.find do |q| 
    q_accumulate += q.dig(:weight)
    q_accumulate >= q_rand
  end

  pipe = pipe_to_process.dig(:pipe)
  
  unless beanstalk.tubes.find(pipe).peek(:ready).nil?
    puts "Getting job from #{pipe}"
    beanstalk.tubes.watch(pipe)
    job = beanstalk.tubes.reserve(1)
    puts "Got job: #{job.id} : tube: #{pipe}"
    job.release delay: 5
  else
    puts "No jobs in #{pipe}"
  end
end
```

Not ideal to wait on a queue that might be empty ( if running multiple processors ).  Any thoughts on better ways to get weighted queues?

**Comment by dkam** [comment]
I wanted to test if that algorithm actually worked and distributed calls to the pipes - this was my test script:

```ruby
#!/usr/bin/env ruby

pipes = [{pipe: 'low_priority', weight: 1}, {pipe: 'medium_priority', weight: 2}, {pipe: 'high_priority', weight: 4}]

q_min = pipes.min_by {|q| q.dig(:weight)}.dig(:weight)
q_max = pipes.inject(0) {|r, q| r + q.dig(:weight) }
range = q_min..q_max

result = { 'low_priority' => 0, 'medium_priority' => 0, 'high_priority' => 0 }

loop do
  q_rand = Random.rand(range)
  q_accumulate = 0
  pipe_to_process = pipes.find do |q| 
    q_accumulate += q.dig(:weight)
    q_accumulate >= q_rand
  end

  pipe = pipe_to_process.dig(:pipe)
  
  result[pipe] += 1

  puts result.map {|p| p.join(': ') }.join(" | ")
end
```

I killed it at 1,000,000 for low priority: 

`low_priority: 1000000 | medium_priority: 2002152 | high_priority: 4004216`

**Comment by dkam** [comment]
Here's an implementation of weighted queues using a layered scheduling approach.

The first layer watches all tubes, so when there's no work available we simply block here — zero wasted traffic on the beanstalkd connection. Each subsequent layer strips out the lower-weighted tubes, narrowing down until only the highest-weighted tubes remain. The number of jobs taken at each layer is proportional to the weights, so higher-weighted tubes get serviced more often while lower-weighted tubes are still guaranteed a share of every cycle.

It’s worth pointing out that if the weight distribution is large, you may get long cycles that a purely random queue selection would avoid.

```ruby
# frozen_string_literal: true

class Beaneater
  # Provides weighted tube selection for reserving jobs.
  #
  # Instead of beanstalkd's default round-robin across watched tubes,
  # this class lets you assign weights to tubes so higher-priority
  # tubes are serviced proportionally more often.
  #
  # == Algorithm: Layered Deterministic Scheduling
  #
  # Given weights { a: 1, b: 2, c: 4 }, the reserver builds a schedule
  # of "layers", peeling off one weight tier at a time:
  #
  #   Layer 1: watch(a, b, c), reserve up to 3 jobs (blocking on first)
  #   Layer 2: watch(b, c),    reserve(0) up to 2 jobs
  #   Layer 3: watch(c),       reserve(0) up to 2 jobs
  #   Repeat from Layer 1.
  #
  # This produces exact weight ratios (1:2:4) when all tubes are full.
  # Layer 1 always watches all tubes, so it doubles as the idle wait —
  # no busy-polling when the system is quiet.
  #
  # @example
  #   client = Beaneater.new('localhost')
  #   reserver = Beaneater::WeightedReserver.new(client, "high" => 4, "medium" => 2, "low" => 1)
  #   loop { reserver.reserve { |job| process(job) } }
  #
  class WeightedReserver
    attr_reader :client, :weights

    # @param client [Beaneater] A beaneater client instance
    # @param weights [Hash{String => Integer}] Tube names mapped to their weights
    def initialize(client, weights)
      raise ArgumentError, "weights must be a non-empty Hash" unless weights.is_a?(Hash) && !weights.empty?

      @client = client
      @weights = weights.transform_keys(&:to_s)
      validate_weights!(@weights)
      @schedule = build_schedule(@weights)
    end

    # Reserve a job using layered deterministic scheduling.
    #
    # Iterates through the schedule layers. The first layer watches all
    # tubes and blocks until a job is available (no busy-polling). Subsequent
    # layers use instant timeouts and only watch higher-weighted tubes.
    #
    # If a layer times out (tubes are empty), it moves to the next layer.
    # After completing all layers, the cycle restarts from layer 1.
    #
    # @yield [job] Optional block called with the reserved job
    # @return [Beaneater::Job] Reserved beaneater job
    def reserve(&block)
      @schedule.each_with_index do |layer, index|
        client.tubes.watch!(*layer[:tubes])
        jobs_to_take = layer[:jobs]

        jobs_to_take.times do |i|
          begin
            if index == 0
              # Layer 1: blocking reserve — all tubes watched, no starvation risk
              job = client.tubes.reserve
            else
              # Subsequent layers: instant timeout — only higher-weight tubes
              job = client.tubes.reserve(0)
            end
            block.call(job) if block
            return job
          rescue Beaneater::TimedOutError
            # Tube(s) empty at this layer, move to next layer
            break
          end
        end
      end

      # All layers exhausted in this cycle — restart with blocking reserve
      # on all tubes. This handles the edge case where layers 2+ had work
      # but layer 1's quota was already filled in a previous call.
      client.tubes.watch!(*@schedule.first[:tubes])
      job = client.tubes.reserve
      block.call(job) if block
      job
    end

    # Returns the computed schedule for inspection/testing.
    #
    # @return [Array<Hash>] Array of { tubes: [...], jobs: N } layers
    def schedule
      @schedule
    end

    private

    # Build the layered schedule from a weights hash.
    #
    # Sorts tubes by weight ascending, then peels off layers. Each layer
    # includes all tubes at or above the current weight tier. The number
    # of jobs per layer is: (current_min_weight) * (number of tubes in layer).
    #
    # @example
    #   build_schedule("a" => 1, "b" => 2, "c" => 4)
    #   # => [
    #   #   { tubes: ["a", "b", "c"], jobs: 3 },  # 1 * 3 tubes
    #   #   { tubes: ["b", "c"],      jobs: 2 },  # 1 * 2 tubes
    #   #   { tubes: ["c"],           jobs: 2 },  # 2 * 1 tube
    #   # ]
    #
    def validate_weights!(weights)
      weights.each do |name, weight|
        unless weight.is_a?(Integer) && weight > 0
          raise ArgumentError, "weight for '#{name}' must be a positive integer, got #{weight.inspect}"
        end
      end
    end

    def build_schedule(weights)
      sorted = weights.sort_by { |_, w| w }
      layers = []
      previous_weight = 0

      sorted.each_with_index do |(_, weight), index|
        remaining_tubes = sorted[index..].map(&:first)
        layer_weight = weight - previous_weight

        if layer_weight > 0
          layers << {
            tubes: remaining_tubes,
            jobs: layer_weight * remaining_tubes.size
          }
        end

        previous_weight = weight
      end

      layers
    end
  end
end
```

**Comment by nesquena** [comment]
If someone can review this and give it a thumbs up I'd be more than happy to merge! 

Thanks for contributing!

**Comment by dkam** [comment]
Mmm - it has a degenerative case - frequent, short, low priority jobs would saturate the lower layer I think, inverting the desired outcome - because beanstalkd delivers jobs FIFO - the lower layer would much more often select “tube a” in the above example.  I’m working on a more probabilistic solution, but it’ll be more chatty talking to beanstalkd.

**Comment by dkam** [comment]
This one randomly selects tubes, based on the weight, skips over empty tubes, and will block when all tubes are empty: 

```ruby
# frozen_string_literal: true

class Beaneater
  # Provides weighted tube selection for reserving jobs.
  #
  # Instead of beanstalkd's default behaviour of returning the earliest
  # ready job across all watched tubes, this class selects a tube based
  # on weighted probability before reserving, so higher-weighted tubes
  # are serviced proportionally more often.
  #
  # When all tubes are empty, falls back to a blocking reserve on all
  # tubes — no busy-polling.
  #
  # @example
  #   client = Beaneater.new('localhost')
  #   reserver = Beaneater::WeightedReserver.new(client, "high" => 4, "medium" => 2, "low" => 1)
  #   loop { reserver.reserve { |job| process(job) } }
  #
  class WeightedReserver
    attr_reader :client, :weights

    # @param client [Beaneater] A beaneater client instance
    # @param weights [Hash{String => Integer}] Tube names mapped to their weights
    def initialize(client, weights)
      raise ArgumentError, "weights must be a non-empty Hash" unless weights.is_a?(Hash) && !weights.empty?

      @client = client
      @weights = weights.transform_keys(&:to_s)
      validate_weights!(@weights)
    end

    # Reserve a job using weighted random tube selection.
    #
    # Picks tubes randomly with probability proportional to their weights.
    # If the selected tube is empty, it's removed from candidates and another
    # is picked. When all tubes are empty, falls back to a blocking reserve
    # on all tubes.
    #
    # @param [Integer] timeout Number of seconds before timing out on the
    #   fallback reserve. nil blocks forever, matching Tubes#reserve behaviour.
    # @yield [job] Optional block called with the reserved job
    # @return [Beaneater::Job] Reserved beaneater job
    def reserve(timeout = nil, &block)
      candidates = @weights.dup

      while candidates.any?
        tube = weighted_select(candidates)
        client.tubes.watch!(tube)
        begin
          job = client.tubes.reserve(0)
          block.call(job) if block
          return job
        rescue Beaneater::TimedOutError
          candidates.delete(tube)
        end
      end

      # All tubes empty — block on all tubes until a job arrives
      client.tubes.watch!(*@weights.keys)
      job = client.tubes.reserve(timeout)
      block.call(job) if block
      job
    end

    private

    def validate_weights!(weights)
      weights.each do |name, weight|
        unless weight.is_a?(Integer) && weight > 0
          raise ArgumentError, "weight for '#{name}' must be a positive integer, got #{weight.inspect}"
        end
      end
    end

    def weighted_select(weights)
      total = weights.values.sum
      r = rand(total)
      weights.each do |tube, weight|
        r -= weight
        return tube if r < 0
      end
      weights.keys.last
    end
  end
end

```

**Comment by dkam** [comment]
I had crack ( well, Claude had a crack ) at adding the functionality to [beanstalkd](https://github.com/dkam/beanstalkd/tree/weighted-tube-selection) directly - it builds and runs.  

I also created a branch for beaneater with the random tube selection system above : https://github.com/dkam/beaneater/tree/weighted-reserver 

Here’s an example with the beanstalkd changes: 

```ruby
 # Small sample (30 jobs) — verify weighted mode is not FIFO
  require "beaneater"

  client = Beaneater.new("localhost:11302")

  20.times { client.tubes.find("high").put "high job" }
  20.times { client.tubes.find("medium").put "medium job" }
  20.times { client.tubes.find("low").put "low job" }

  client.connection.transmit("watch high 4")
  client.connection.transmit("watch medium 2")
  client.connection.transmit("watch low 1")
  client.connection.transmit("reserve-mode weighted")

  counts = Hash.new(0)
  30.times do
    job = client.tubes.reserve(0)
    counts[job.tube] += 1
    job.delete
  end

  puts counts.sort_by { |k, v| -v }.map { |k, v| "#{k}: #{v}" }.join(" | ")
  # Example output: high: 14 | medium: 11 | low: 5

  client.close
```

This outputs `high: 17 | medium: 10 | low: 3`

```ruby
  # Larger sample (300 jobs) — verify distribution converges to 4:2:1 weights
  require "beaneater"

  client = Beaneater.new("localhost:11302")

  200.times { client.tubes.find("high").put "high" }
  200.times { client.tubes.find("medium").put "medium" }
  200.times { client.tubes.find("low").put "low" }

  client.connection.transmit("watch high 4")
  client.connection.transmit("watch medium 2")
  client.connection.transmit("watch low 1")
  client.connection.transmit("reserve-mode weighted")

  counts = Hash.new(0)
  300.times do
    job = client.tubes.reserve(0)
    counts[job.tube] += 1
    job.delete
  end

  total = counts.values.sum.to_f
  puts counts.sort_by { |k, v| -v }.map { |k, v| "#{k}: #{v} (#{(v / total * 100).round(1)}%)" }.join(" | ")
  puts "Expected ratio: high 57.1% | medium 28.6% | low 14.3%"
  # Example output: high: 174 (58.0%) | medium: 79 (26.3%) | low: 47 (15.7%)

  client.close
```

```ruby
high: 180 (60.0%) | medium: 72 (24.0%) | low: 48 (16.0%)
Expected ratio: high 57.1% | medium 28.6% | low 14.3%
```
